pkg/templatelookup/regular.go

```diff
@@ -52,53 +52,51 @@ func (t *TemplateLookup) GetRegularTemplates(ctx context.Context, kyma *v1beta2.
 			continue
 		}
 		template := t.GetAndValidate(ctx, module.Name, module.Channel, kyma.Spec.Channel)
-		if template.Err != nil {
-			templates[module.Name] = &template
-			continue
-		}
-		if err := t.descriptorProvider.Add(template.ModuleTemplate); err != nil {
-			template.Err = fmt.Errorf("failed to get descriptor: %w", err)
-		}
-
-		templates[module.Name] = &template
-	}
-
-	for moduleName, moduleTemplate := range templates {
-		if moduleTemplate.Err != nil {
-			continue
-		}
-
-		if moduleTemplate.IsInternal() && !kyma.IsInternal() {
-			moduleTemplate.Err = fmt.Errorf("%w: internal module", ErrTemplateNotAllowed)
-			templates[moduleName] = moduleTemplate
-		}
-		if moduleTemplate.IsBeta() && !kyma.IsBeta() {
-			moduleTemplate.Err = fmt.Errorf("%w: beta module", ErrTemplateNotAllowed)
-			templates[moduleName] = moduleTemplate
-		}
-	}
-
-	for moduleName, moduleTemplate := range templates {
-		template := moduleTemplate
+		templates[module.Name] = FilterTemplate(template, kyma, t.descriptorProvider)
 		for i := range kyma.Status.Modules {
 			moduleStatus := &kyma.Status.Modules[i]
-			if moduleMatch(moduleStatus, moduleName) && template.ModuleTemplate != nil {
-				t.checkValidTemplateUpdate(ctx, template, moduleStatus)
+			if moduleMatch(moduleStatus, module.Name) {
+				descriptor, err := t.descriptorProvider.GetDescriptor(template.ModuleTemplate)
+				if err != nil {
+					msg := "could not handle channel skew as descriptor from template cannot be fetched"
+					template.Err = fmt.Errorf("%w: %s", ErrTemplateUpdateNotAllowed, msg)
+					continue
+				}
+				MarkInvalidChannelSkewUpdate(ctx, &template, moduleStatus, descriptor.Version)
 			}
 		}
-		templates[moduleName] = template
+		templates[module.Name] = &template
 	}
-
 	return templates
 }
 
+func FilterTemplate(template ModuleTemplateInfo, kyma *v1beta2.Kyma,
```

Comment:
1. My first attempt is try to not make this function use TemplateLookup as receiver, because, it's not necessary to introduce client.Reader as intermediate dependencies, however, actually for this function, even not necessary to introduce DescriptorProvider.
2. renamed to `ValidateTemplateMode`
3. as the reason in 1